corehq/ex-submodules/couchforms/models.py

```diff
@@ -349,24 +349,47 @@ def xml_md5(self):
     def archive(self, user_id=None):
         if self.is_archived:
             return
-        self.doc_type = "XFormArchived"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='archive',
-        ))
-        self.save()
-        xform_archived.send(sender="couchforms", xform=self)
+        # If this archive was initiated by a user, delete all other stubs for this action so that this action
+        # isn't overridden
+        from couchforms.models import UnfinishedArchiveStub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True) as archive_stub:
+            self.doc_type = "XFormArchived"
+
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='archive',
+            ))
+            self.save()
+            archive_stub.archive_history_updated()
+            xform_archived.send(sender="couchforms", xform=self)
 
     def unarchive(self, user_id=None):
         if not self.is_archived:
             return
-        self.doc_type = "XFormInstance"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='unarchive',
-        ))
-        XFormInstance.save(self)  # subclasses explicitly set the doc type so force regular save
-        xform_unarchived.send(sender="couchforms", xform=self)
+        # If this unarchive was initiated by a user, delete all other stubs for this action so that this action
+        # isn't overridden
+        from couchforms.models import UnfinishedArchiveStub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False) as archive_stub:
+            self.doc_type = "XFormInstance"
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='unarchive',
+            ))
+            XFormInstance.save(self)  # subclasses explicitly set the doc type so force regular save
+            archive_stub.archive_history_updated()
+            xform_unarchived.send(sender="couchforms", xform=self)
+
+    def publish_archive_action_to_kafka(self, user_id, archive):
+        from couchforms.models import UnfinishedArchiveStub
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        # Delete the original stub
+        UnfinishedArchiveStub.objects.filter(xform_id=self.form_id).all().delete()
+        with unfinished_archive(instance=self, user_id=user_id, archive=archive):
+            xform_archived.send(sender="couchforms", xform=self)
```

Comment:
this should be `xform_unarchived` when `archive == False`

Reply:
https://github.com/dimagi/commcare-hq/pull/22470/commits/d393107933168eeeb3f25f3319cb8d25e25bdbad